tests/integration/test_envelope.py

```diff
@@ -579,3 +580,52 @@ def test_sample_rates_metrics(mini_sentry, relay_with_processing, events_consume
 
     event, _ = events_consumer.get_event()
     assert event["_metrics"]["sample_rates"] == sample_rates
+
+
+def test_buffer_envelopes_without_global_config(
+    mini_sentry, relay_with_processing, events_consumer
+):
+    """
+    Checks that we buffer envelopes until we have a global config in processing mode.
+    """
+
+    include_global = False
+    original_endpoint = mini_sentry.app.view_functions["get_project_config"]
+
+    @mini_sentry.app.endpoint("get_project_config")
+    def get_project_config():
+        nonlocal include_global
+
+        res = original_endpoint().get_json()
+        if not include_global:
+            res["global"] = None
+        return res
+
+    mini_sentry.add_basic_project_config(42)
+    events_consumer = events_consumer()
+    options = {"cache": {"global_config_fetch_interval": 1}}
+    relay = relay_with_processing(options=options)
+
+    envelope_qty = 5
+    for _ in range(envelope_qty):
+        envelope = Envelope()
+        envelope.add_event({"message": "hello, world!"})
+        relay.send_envelope(42, envelope)
+
+    time.sleep(2)
+    events_consumer.assert_empty()
+
+    include_global = True
+    # Clear errors because we log error when we request global config yet we dont receive it.
+    assert len(mini_sentry.test_failures) > 0
+    for err in mini_sentry.test_failures:
+        assert (
+            str(err[1])
+            == "Relay sent us event: global config missing in upstream response"
+        )
+    mini_sentry.test_failures.clear()
+
+    # Check that we received exactly {envelope_qty} envelopes.
+    for _ in range(envelope_qty):
+        events_consumer.get_event(timeout=2)
+    events_consumer.assert_empty()
```

Comment:
this one doesn't require sleeping first, since we assert that it's not empty in the previous lines, and we just wanna check that we take out exactly `envelope_qty` envelopes